Feature/acr2 pipeline

caldwellst opened the pull request:
Setting up a draft PR so we have a discussion place! You can see the files I've changed and a quick example file in `exploration` intended to showcase its usage. There's a bit more to do of course, but should go much faster now that I've schooled myself in everything NetCDF/xarray/rioxarray!

hannahker commented on src/utils_general/raster_manipulation.py:
```diff
@@ -152,7 +152,20 @@ def compute_raster_statistics(
+            da_clip = da_clip.rio.clip(
+                gdf_adm[geom_col], all_touched=all_touched
+            )
+        except Exception:
```
As you mentioned in your TODO, would be good to catch the specific exception here.

hannahker commented on src/indicators/drought/arc2_precipitation.py:
```diff
@@ -23,103 +31,194 @@
     ):
         self.country_iso3 = country_iso3
+
+        if not isinstance(date_min, date):
```
Ah, good to do this!

caldwellst commented on src/indicators/drought/arc2_precipitation.py:
```diff
@@ -0,0 +1,609 @@
+            f"https://iridl.ldeo.columbia.edu/SOURCES/.NOAA/.NCEP/.CPC/.FEWS/"
+            f".Africa/.DAILY/.ARC2/.daily/.est_prcp/T/"
+            f"%28{date_min_dl:%-d%%20%b%%20%Y}%29"
+            f"%28{date_max_dl:%-d%%20%b%%20%Y}%29RANGEEDGES/"
```
Oooh interesting! Good catch (and sorry that was left to you!). Have replaced `%-d` with `date_min_dl.day` and should now be a cross platform solution since that was the only platform dependent formatter there. Just pushed solution if you want to test.

Tinkaa commented on src/indicators/drought/arc2_precipitation.py:
```diff
@@ -0,0 +1,710 @@
+
+            master_merge = xr.concat([master, raw], dim="T")
+
+            os.remove(raw_filepath)
```
This is something I learnt from Monica recently. You always prefer not to remove files as it increases the risk that you accidentally remove something that you don't want to remove. A solution is to instead use a temporary file/folder. I am not too sure what the best structure would be here, but what I did for FN is [here](https://github.com/OCHA-DAP/pa-aa-toolbox/blob/9390ea80a8d2dc5348e04611f5cb4ba45b0fc2f6/src/aatoolbox/datasources/fewsnet/fewsnet.py#L55). Monica might have better advice

caldwellst replied:
Good idea. have added as TODO for transition to `pa-aa-toolbox`.

Tinkaa commented on src/indicators/drought/arc2_precipitation.py:
```diff
@@ -0,0 +1,710 @@
+            os.remove(raw_filepath)
+            os.remove(master_filepath)
+
+            if "_FillValue" in master_merge.encoding and np.isnan(
```
Add comment why you are doing this

Tinkaa commented on src/indicators/drought/arc2_precipitation.py:
```diff
@@ -0,0 +1,710 @@
+
+            master_merge.to_netcdf(master_filepath)
+
+        return
```
I don't know what official best practice is, but [discussions](https://stackoverflow.com/questions/13307158/most-pythonic-way-of-function-with-no-return) seem to converge to not add `return` if it never returns anything

Tinkaa commented on src/indicators/drought/arc2_precipitation.py:
```diff
@@ -0,0 +1,710 @@
+            f"{self.range_x[0]}_{self.range_x[1]}_"
+            f"{self.range_y[0]}_{self.range_y[1]}_{end}.nc"
+        )
+        return directory / Path(filename)
```
Once the first part i.e. `directory` is a path, you don't have to convert the string into a Path before concatenating

Tinkaa commented on src/indicators/drought/arc2_precipitation.py:
```diff
@@ -0,0 +1,710 @@
+        try:
+            with rioxarray.open_rasterio(raw_filepath, masked=True) as ds:
+                da = ds.load()
+                da = da.squeeze().rio.write_crs("EPSG:4326")
```
maybe set the crs as a global constant?

Tinkaa commented on src/indicators/drought/arc2_precipitation.py:
```diff
@@ -0,0 +1,710 @@
+
+            loaded_dates = mr.indexes[
+                "T"
+            ].to_datetimeindex() - pd.to_timedelta("12:00:00")
```
add comment why you extract 12 hours

Tinkaa commented on src/indicators/drought/arc2_precipitation.py:
```diff
@@ -0,0 +1,710 @@
+        master.sortby("T").to_netcdf(master_filepath)
+
+
+class DrySpells(ARC2):
```
Would it be an idea to move the dryspells to a separate file? And maybe don't make it dependent on ARC2? Such that other datasets such as CHIRPS could also be given as input (at first glance code doesn't seem ARC2 specific? )

caldwellst replied:
Yes, definitely the idea, will be part of refactor into toolbox!

Tinkaa commented on src/mwi/get_arc2_data.py:
```diff
@@ -0,0 +1,55 @@
+import os
```
As discussed, this file is outdated so remove

Tinkaa replied:
@caldwellst you resolved this one, but I still think it can be deleted?

Tinkaa commented on analyses/mwi/dryspells_trigger/exploration/test_arc2_processing.md:
```diff
@@ -0,0 +1,92 @@
+    range_y = ("20S", "5S")
+)
+
+arc2_test.download_data(master=True)
```
If I change this to False and set the date range to one that is not yet present in the master file, I get the error `TypeError: unhashable type: 'list'`

Tinkaa commented on requirements.txt:
```diff
@@ -44,4 +44,4 @@ nbconvert==5.6.1
 plotly==5.1.0
 dataframe-image==0.1.1
-
+h5netcdf==0.11.0
```
Why did you add this one? Cause I don't see it used anywhere or am I missing something?

Tinkaa replied:
@caldwellst why do we need this?

caldwellst commented on src/indicators/drought/arc2_precipitation.py:
```diff
@@ -0,0 +1,710 @@
+        )
+
+        # TODO: explore specific errors generated by request
+        try:
```
No, wasn't very familiar with wider functions available at that time! Will add in once we move to `pa-aa-toolbox`.

caldwellst commented on src/indicators/drought/arc2_precipitation.py:
```diff
@@ -0,0 +1,710 @@
+            with xr.open_dataarray(master_filepath) as ds:
+                master = ds.load()
+            with xr.open_dataarray(raw_filepath) as ds:
+                raw = ds.load()
```
Ha yeah, I think I saw this elsewhere in someone's code and kept it. Can be refactored eventually.

caldwellst commented:
Thank you @hannahker (hello from the beyond!) and @Tinkaa for the great comments. I've finally gone through and addressed all of them. Some of them I have added as TODO for when we bring this into the toolbox, but I think for now should be good enough for merging into the `pa-anticipatory-action` repo at least. Feel free to let me know what you think. As well, the `15_mwi_obs_trigger_monitoring` notebook should be sufficient to explain how to do the monitoring and generate the graphs as needed, although there's additional functionality available in the `src` code.

Tinkaa commented on src/indicators/drought/arc2_precipitation.py:
```diff
@@ -0,0 +1,884 @@
+            right=total_df,
+            on=[adm_col, "ds_confirmation"],
+            how="left",
+        )
```
I was trying to comeup with an easier method to compute the total rainfall but couldn't so this looks good!

Tinkaa commented on src/indicators/drought/arc2_precipitation.py:
```diff
@@ -0,0 +1,884 @@
+            ]
+        return df
+
+    def _write_to_monitoring_file(self, dry_spells: Union[None, int] = None):
```
this function seems unused and `_get_monitoring_filepath` undefined? Fine for now, just add acomment

Tinkaa commented on src/indicators/drought/arc2_precipitation.py:
```diff
@@ -0,0 +1,884 @@
+        )
+        return directory / filename
+
+    def load_aggregated_data(self, filter: bool = False) -> pd.DataFrame:
```
I think filter is also a built-in name so better to choose another name

Tinkaa commented on src/indicators/drought/arc2_precipitation.py:
```diff
@@ -0,0 +1,884 @@
+            da = self.load_raw_data()
+            da_date = da.indexes["T"].to_datetimeindex().date
+            da = da[
+                (da_date >= self.date_min) & (da_date <= self.date_max), :, :
```
you can also do this with `sel`

Tinkaa commented on src/indicators/drought/arc2_precipitation.py:
```diff
@@ -0,0 +1,884 @@
+        df = self.load_dry_spell_data(filter=filter)
+        return len(np.unique(df.iloc[:, 0]))
+
+    def days_under_threshold(self, raster: bool = True) -> pd.DataFrame:
```
nice function!

Tinkaa commented on requirements.txt:
```diff
@@ -44,5 +44,6 @@ nbconvert==5.6.1
 pre-commit==2.13.0
 plotly==5.1.0
 dataframe-image==0.1.1
+h5netcdf==0.11.0
```
my pycharm is indicating that `fiona` is imported in `arc2_precipitation.py` but not included in `requirements.txt`. Should it be included?

caldwellst replied:
I didn't put in there because it's a requirement of `geopandas` so should already be there (added because it allows excepting the specific errors from `gpd.read_file`), but good idea to be explicit (should be fine anyway when porting to toolbox).

Tinkaa commented on analyses/mwi/dryspells_trigger/exploration/test_arc2_processing.md:
```diff
@@ -0,0 +1,96 @@
+## ARC2 processing
```
I think that we can also remove this file now that we have `15_mwi_obs_trigger_monitoring.md` ?

caldwellst replied:
Yeah, but might just leave now anyway, find it useful for testing as I've been amending.

Tinkaa commented on analyses/mwi/dryspells_trigger/exploration/test_resampling.md:
````diff
@@ -0,0 +1,270 @@
+```python
````
I would rename the file to smth like arc2_aggregation_methods

Tinkaa commented on analyses/mwi/dryspells_trigger/exploration/test_resampling.md:
````diff
@@ -0,0 +1,270 @@
+
+```python
+
+arc2_approx = DrySpells(
````
this is old code that should be removed? getting an error

Tinkaa commented on analyses/mwi/dryspells_trigger/exploration/mwi_agg_exploration.Rmd:
````diff
@@ -0,0 +1,382 @@
+```{r setup, include=FALSE}
+knitr::opts_chunk$set(echo = FALSE)
+
+library(tidyverse)
````
I had to do `install.packages("Rcpp")`. Not sure if that should be defined in the proj file or so?

caldwellst replied:
No, would be if we were using something like `renv` but I think that's overkill, should be all good now!

Tinkaa commented on analyses/mwi/dryspells_trigger/exploration/mwi_agg_exploration.Rmd:
```diff
@@ -0,0 +1,382 @@
+methods <- c("approximate_mask", "centroid", "touching")
+
+df_list <- map(methods, ~read_csv(
+  file.path(arc2_dir, glue("arc2_{.x}_mwi_dry_spells_14_days_32E_36E_20S_5S_master.csv"))
```
getting an error that this file doesn't exist

Tinkaa replied:
Didn't feel like debugging right now so didn't review the rest of the file yet..

Tinkaa replied:
And for some reason it is stating that I don't have permission to open the html file. So let me know when you debugged this code or the html and I can review

caldwellst replied:
Are you still getting errors here @Tinkaa? I get none and the fine does exist on my machine.

caldwellst replied:
Ha, it was just because I'd tried to be more culturally sensitive and change master to main as Hannah suggested, forgot to change this!

Tinkaa commented on analyses/mwi/dryspells_trigger/exploration/test_resampling.md:
```diff
@@ -0,0 +1,270 @@
+    "percent_area":xr.where(ds_rolling <= 2, 1, 0).mean(dim=["x", "y"]) * 100
+})
+
+df.to_csv(
```
comment out file saving. 

And maybe save to `exploration` instead of `processed`?

Tinkaa commented on analyses/mwi/dryspells_trigger/exploration/test_resampling.md:
````diff
@@ -0,0 +1,270 @@
+```
+
+```python
+arc2_centr.dry_spell_count()
````
you renamed this to `count_dry_spells`

Tinkaa commented on analyses/mwi/dryspells_trigger/exploration/test_resampling.md:
````diff
@@ -0,0 +1,270 @@
+
+```python
+df = arc2_centr.days_under_threshold()
+gdf_merged = gdf_adm2.merge(df, on="ADM2_PCODE")
````
NameError: name 'gdf_adm2' is not defined

Tinkaa commented on analyses/mwi/dryspells_trigger/exploration/test_resampling.md:
````diff
@@ -0,0 +1,270 @@
+```python
+%load_ext autoreload
````
I am unsure what the diff between this file and `mwi_agg_exploration.Rmd` is. Explain a bit and fix errors, then I can review again

Tinkaa commented on analyses/mwi/dryspells_trigger/15_mwi_obs_trigger_monitoring.md:
````diff
@@ -0,0 +1,154 @@
+So, if we have triggered, we also want to plot the cumulative rainfall across our area of interest and monitoring period, as well as numbers of days without rain above 2mm. We can plot with any of the methods of aggregation, but chose to use centroid for below.
+
+```python
+# get number of consecutive days under threshold per area
````
I think the graphics can be improved a bit. But assume that is not the focus of this PR and we can discuss later

Tinkaa commented on analyses/mwi/dryspells_trigger/15_mwi_obs_trigger_monitoring.md:
````diff
@@ -0,0 +1,154 @@
+    f"Touching method: {arc2_touch.count_days_under_threshold(DAYS)}\n"
+    f"Approx m method: {arc2_approx.count_days_under_threshold(DAYS)}"
+)
+```
````
not super important but might be nice to add the max number of cons days observed in an area? 

`print(
    f"Centroid method: {arc2_centr.days_under_threshold(raster=False).days_under_threshold.max()}\n"
    f"Touching method: {arc2_touch.days_under_threshold(raster=False).days_under_threshold.max()}\n"
    f"Approx m method: {arc2_approx.days_under_threshold(raster=False).days_under_threshold.max()}"
)`

Tinkaa commented on src/indicators/drought/arc2_precipitation.py:
```diff
@@ -0,0 +1,884 @@
+        """
+        t_col = df.columns[0]
+        precip_col = df.columns[1]
+        adm_col = df.columns[2]
```
this occurs in more places so won't comment on it anymore but yes wouldn't recommend this method

Tinkaa reviewed (approved):
Oh man, Seth I am so sorry for this massive amount of comments. It creates a very skewed picture cause I think you did a great job!! I did a thorough review of all computations and I think they are all correct. So it is really only comments on the code structure and building in some checks. 

This means the code is ready to be used for monitoring as well :partying_face: Thanks for all of this!

Tinkaa commented on analyses/mwi/dryspells_trigger/15_mwi_obs_trigger_monitoring.md:
````diff
@@ -0,0 +1,154 @@
+
+```python
+arc2_centr.aggregate_data(POLY_PATH, "ADM2_PCODE")
+arc2_centr.calculate_rolling_sum()
````
I was just trying to run the notebook and got an error here: `first argument must be an iterable of pandas objects, you passed an object of type "DataFrame"` 
According to [SO](https://stackoverflow.com/questions/39534676/typeerror-first-argument-must-be-an-iterable-of-pandas-objects-you-passed-an-o/50459002) had to make a list object when concatenating in `arc2_precipitation.py` in line 582 and 594. That did solve the problem for me

caldwellst replied:
In other work got rid of that code, so no worries now as it all runs ha!

Tinkaa commented on analyses/mwi/dryspells_trigger/15_mwi_obs_trigger_monitoring.md:
```diff
@@ -0,0 +1,154 @@
+da_cum.plot(ax = ax,
+            cmap='Greys')
+gdf_adm3.plot(ax=ax, facecolor="none", alpha=0.5)
+plt.title(f"Cumulative rainfall from {arc2_centr.monitoring_start}")
```
Here you are indicating it is the cumulative rainfall from `monitoring_start`. But looking in the code of `cumulative_rainfall()` it is the cum rainfall from `date_min` where `date_min=monitoring_start-rolling_window`. So should at least be aligned. I think it would make most sense to set the start date of `cumulative_rainfall()` to `monitoring_start` instead of `date_min`

caldwellst replied:
Very good catch on the mislabeling, missed it in the last effort code up prior to leave! Here, I think we should still look from `date_min`, because that is when we start tracking for dry spells (that's the first day included in a cumulative sum).

Tinkaa commented on src/indicators/drought/arc2_precipitation.py:
```diff
@@ -0,0 +1,884 @@
+
+        :param raw_filepath: Path to raw file to load. If `None`,
+            loads main file.
+        """
```
I don't know if this is the best place to do so, but would it be an idea to somewhere add `self` parameters of the data availability between the min and max date. So e.g. if we set the `date_max` to `2022-01-04` but there is only data till `2022-01-01`, it would be helpful to be able to easily retrieve that somehow.

caldwellst replied:
Love the idea.

caldwellst commented on src/indicators/drought/arc2_precipitation.py:
```diff
@@ -0,0 +1,884 @@
+            If `True`, looks for missing values through entire
+            dataset and ensures they are re-downloaded. If an
+            ISO 8601 string is passed, then missing values are
+            only looked for on or after that date.
```
My understanding was no, but that's probaby a good option to have anyway. Will look into it when refactoring to toolbox!

caldwellst commented on src/indicators/drought/arc2_precipitation.py:
```diff
@@ -0,0 +1,884 @@
+        try:
+            with rioxarray.open_rasterio(raw_filepath, masked=True) as ds:
+                da = ds.load()
+                da = da.squeeze().rio.write_crs(ARC2_CRS)
```
Didn't here because it's squeezed and returning the array and explicitly resaving it.

caldwellst commented on src/indicators/drought/arc2_precipitation.py:
```diff
@@ -0,0 +1,884 @@
+        available in the main file, and then merged
+        into the main file.
+
+        :param main: Boolean on whether to set download as
```
Yeah, would definitely move to the clobber standard we've done elsewhere, would make it much simpler. This was when I was first just trying something out and then went haywire on it. Will shift to clobber once we drop into the main toolbox. This would just enable us to do the last option that you're talking about to keep it simple, which I really like.

caldwellst commented on src/indicators/drought/arc2_precipitation.py:
```diff
@@ -0,0 +1,884 @@
+                    self._download(
+                        date_min_dl=dates[0],
+                        date_max_dl=dates[1],
+                        main=False,
```
In this case, we don't want that, just want those downloaded and merged into main.

caldwellst commented on src/indicators/drought/arc2_precipitation.py:
```diff
@@ -0,0 +1,884 @@
+        :param main: Boolean on whether to set download as
+            main if `True`, or only download missing data
+            and merge to main if `False`.
+        :param replace_missing: Only relevant if not `main`.
```
I think we should just get rid of for simplicity. The issue was just that there are dates that lack data within the API that you wouldn't necessarily want to re-download, however, should probably just streamline it out using clobber.

caldwellst commented on src/indicators/drought/arc2_precipitation.py:
```diff
@@ -0,0 +1,884 @@
+                        main=False,
+                    )
+
+                self._sort_raw_data()
```
Hmm, I think will leave as is just for the moment, don't need to sort every time we load, just whenever new data is downloaded and added to `main`.

caldwellst commented on src/indicators/drought/arc2_precipitation.py:
```diff
@@ -0,0 +1,884 @@
+                self._sort_raw_data()
+            else:
+                logger.info("No additional data needs downloading.")
+
```
Gonna drop `main` and move to the `clobber` system for much more simplicity. Will keep this here though for that refactor, because a nice and simple way to approach looking for missing dates and replacing them.

caldwellst commented on src/indicators/drought/arc2_precipitation.py:
```diff
@@ -0,0 +1,884 @@
+        if not isinstance(monitoring_end, date):
+            monitoring_end = date.fromisoformat(monitoring_end)
+
+        self.monitoring_start = monitoring_start
```
Have noted in a TODO for the refactor.

caldwellst commented on src/indicators/drought/arc2_precipitation.py:
```diff
@@ -0,0 +1,884 @@
+        main.sortby("T").to_netcdf(main_filepath)
+
+
+class DrySpells(ARC2):
```
Ha yeah, been thinking about the classes and will definitely be totally overhauling them. I was going buckwild in the class system at the time cause was just really learning about it, and wouldn't take that approach again ha.

caldwellst commented on src/indicators/drought/arc2_precipitation.py:
```diff
@@ -0,0 +1,884 @@
+        polygon_path: Union[Path, str] = None,
+        bound_col: str = None,
+        reprocess: bool = False,
+        redownload: bool = False,
```
Yeah, agreed. Will replace during the refactoring.

caldwellst commented on src/indicators/drought/arc2_precipitation.py:
```diff
@@ -0,0 +1,884 @@
+        :param polygon_path: Path to polygon file for clipping and aggregating
+            raster data.
+        :param bound_col: Column in polygon file to aggregate raster to.
+        :param reprocess: Boolean, if `True` reprocesses all raster data.
```
The idea would be that reprocessing the data could be necessary if somehow the raw data changes, but maybe not worth adding in here because the likelihood of that happening is so low. Accounting for something that is a rare event I guess.

caldwellst commented on src/indicators/drought/arc2_precipitation.py:
```diff
@@ -0,0 +1,884 @@
+            f"arc2_{self.agg_method}_long_{self.country_iso3}_"
+            f"{self.range_x[0]}_{self.range_x[1]}_"
+            f"{self.range_y[0]}_{self.range_y[1]}_"
+            f"main.csv"
```
Nope. To help with testing, I will remove those in the refactor to see if the new system matches up with the one we have now.

caldwellst commented on src/indicators/drought/arc2_precipitation.py:
```diff
@@ -0,0 +1,884 @@
+        with xr.open_dataarray(main_filepath) as ds:
+            main = ds.load()
+
+        main.sortby("T").to_netcdf(main_filepath)
```
Very strange, as have tested before and getting that error, I was able to fix this by adding in the above code you're referencing. Super strange. Let me do some testing, have been exploring for a bit now.

caldwellst replied:
I cannot reproduce your error, but made some changes which I thought might theoretically fix it. Could you provide the code that generated this error?

Tinkaa replied:
Hmmm maybe it is my versioning then.. But I still get the error. Below an example, could you try? (also seems you didn't push your changes yet so maybe they will solve it?) 

``` python
from src.indicators.drought.arc2_precipitation import DrySpells
RANGE_X = ("5E", "10E")
RANGE_Y = ("15S", "10S")

MONITORING_START = "2018-01-07"
MONITORING_END = "2018-03-07"


# centroid method
arc2_centr = DrySpells(
    country_iso3 = "mwi",
    monitoring_start = MONITORING_START,
    monitoring_end = MONITORING_END,
    range_x = RANGE_X,
    range_y = RANGE_Y
)

arc2_centr.download_data(main=True)

MONITORING_START = "2018-04-07"
MONITORING_END = "2018-06-07"


# centroid method
arc2_centr = DrySpells(
    country_iso3 = "mwi",
    monitoring_start = MONITORING_START,
    monitoring_end = MONITORING_END,
    range_x = RANGE_X,
    range_y = RANGE_Y
)

arc2_centr.download_data(main=False)

caldwellst replied:
Really stuck figuring out what this is. Will have to put on backburner to investigate exactly what is happening here, very frustrating. Back to square one again for me.

Hope that simplifying the process in the toolbox will avoid these issues!!!

caldwellst commented on src/indicators/drought/arc2_precipitation.py:
```diff
@@ -0,0 +1,884 @@
+            aggregated_filepath.parent.mkdir(parents=True, exist_ok=True)
+
+        # explicitly remove missing values
+        da.values[da.values == -999] = np.NaN
```
Hmm yes, I need to double check, I think I added before I was correcting the `_FillValue` attribute. Will confirm and fix!

caldwellst commented on src/indicators/drought/arc2_precipitation.py:
```diff
@@ -0,0 +1,884 @@
+            df_zonal_stats[data_col].isna(), True, df_zonal_stats["infilled"]
+        )
+
+        df_zonal_stats[data_col] = df_zonal_stats.groupby(bound_col)[
```
So, `interpolate()` in that case defaults to doing linear interpolation between those dates. However, what we've seen is just about 5 days in total with missing data (something like that), all of them on their own except a single instance of 2 days in a row. I looked at `limit` before but it just restricts the number of `NaN` cells filled. So, if you have 3 in a row and `limit = 1`, it actually fills the first but not the last 2. Have added as TODO to look into something more permanent if we wanted to restrict, but not sure that's an actual problem we'll run into (at least it wasn't in MWI).

caldwellst commented on src/indicators/drought/arc2_precipitation.py:
```diff
@@ -0,0 +1,884 @@
+            t_col = df.columns[0]
+            # first we look at data before our previous roll sum
+            # and recalculate if sufficient data available
+            if min(prev_rollsum_df[t_col]) - min(df[t_col]) >= pd.Timedelta(
```
Yeah, will consider removing the recomputation later on. However, if on the backend in the toolbox, I guess not a huge deal to keep in if code is streamlined/isn't crazy.

caldwellst commented on src/indicators/drought/arc2_precipitation.py:
```diff
@@ -0,0 +1,884 @@
+        Identifies dry spells based on latest rolling
+        sum values.
+
+        :param reprocess: Boolean, if `True` reprocesses
```
Yep, agreed! Will remove in refactor.

caldwellst commented:
Thanks @Tinkaa for the extremely detailed comments. I've gone through them all. However, whereas before I was not seeing at all the issues you raised regarding the `_FillValue` and `missing_value` or the unknown dims, I am now, and struggled all day yesterday to really dig deep and figure out exactly what the issue was. Gotta come back to this, but at least just re-downloading data works as a current solution for the short time being.

Tinkaa commented:
Oh no I am sorry.. Yes definitely leave it for now then. 
Maybe you can even merge this branch and make an issue about it?

caldwellst commented:
Thanks @Tinkaa, have created #262.
